lib/Sema/SemaChecking.cpp

```diff
@@ -10621,6 +10621,64 @@ void Sema::DiagnoseSelfMove(const Expr *LHSExpr, const Expr *RHSExpr,
                                         << RHSExpr->getSourceRange();
 }
 
+
+//===--- CHECK: Checked scope -------------------------===//
+void Sema::DiagnoseCheckedDecl(const DeclStmt *DS) {
+  // Checked C - check consistency between compound stmt checked property
+  // and declaration type
+  DeclGroupRef DG = DS->getDeclGroup();
+  for (DeclGroupRef::const_iterator D = DG.begin(), DEnd = DG.end(); D != DEnd;
+       ++D) {
+    if (const VarDecl *VD = dyn_cast<VarDecl>(*D)) {
+      DiagnoseCheckedDecl(VD);
+    }
+  }
+}
+
+void Sema::DiagnoseCheckedDecl(const DeclaratorDecl *Decl) {
+  // Checked C - check consistency between checked property and declaration
+  // checked pointer type or unchecked pointer type with bounds-safe interface
+  // is only allowed in checked scope or funcion
+  const DeclaratorDecl *TargetDecl = nullptr;
+  int TypeKind;
+  QualType Ty;
+  if (const ParmVarDecl *Parm = dyn_cast<ParmVarDecl>(Decl)) {
+    TargetDecl = Parm;
+    TypeKind = 0; // function param type
+    Ty = Parm->getType();
+  }
+  else if (const FunctionDecl *Func = dyn_cast<FunctionDecl>(Decl)) {
+    TargetDecl = Func;
+    TypeKind = 1; // function return type
+    Ty = Func->getReturnType();
+  }
+  else if (const VarDecl *Var = dyn_cast<VarDecl>(Decl)) {
+    TargetDecl = Var;
+    TypeKind = 2; // decl var type
+    Ty = Var->getType();
+  }
+  else if (const FieldDecl *Field = dyn_cast<FieldDecl>(Decl)) {
+    TargetDecl = Field;
+    TypeKind = 3; // member type
+    Ty = Field->getType();
+  }
+
+  bool checkedScopeError = false;
+  if (TargetDecl) {
+    // type is unchecked pointer w/o bounds-safe interface
+    // type is unchecked pointer & its bounds-safe interface is also unchecked
+    QualType InterOpTy = GetCheckedCInteropType(TargetDecl);
+    if ((Ty->isUncheckedPointerType() || Ty->isUncheckedArrayType()) &&
```

Comment:
We will need a more complex way to determine whether a type uses an unchecked type.   We have to check what the pointer points to and whether that is an unchecked type.  We have a routine in Type.h that determines whether a type mentions a checked type.  We will need a similar routine for a checked type..

I think what you have here is a great first step.   I think adding more complex checking is beyond the scope of this pull request.   I have opened work item #200 to track this additional work.   I do not think you should extend this pull request to add this functionality - it recommend doing it separately.